model/interactive/src/main/proto/beam_interactive_api.proto

```diff
@@ -40,18 +40,22 @@ import "google/protobuf/timestamp.proto";
 message TestStreamFileHeader {
   // The PCollection tag this stream is associated with.
   string tag = 1;
+
+  // The file format version. This is used to ensure backwards compatibility
+  // when decoding from file.
+  int32 version = 2;
 }
 
 // A record is a recorded element that a source produced. Its function is to
 // give enough information to create a faithful recreation of the original
 // stream of data.
 message TestStreamFileRecord {
   oneof recorded_event {
-    // The recorded element with its event timestamp (when it was produced).
-    org.apache.beam.model.pipeline.v1.TestStreamPayload.TimestampedElement element = 1;
+    // The recorded bundle with its event timestamp (when it was produced).
+    org.apache.beam.model.pipeline.v1.TestStreamPayload.Event.AddElements element_event = 1;
 
     // Indicating the output watermark of the source changed.
-    google.protobuf.Timestamp watermark = 2;
+    org.apache.beam.model.pipeline.v1.TestStreamPayload.Event.AdvanceWatermark watermark_event = 2;
   }
 
   // The wall-time timestamp of either the new element or watermark change.
```

Comment:
Why do the recorded events not have AdvanceProcessingTime events?

Reply:
It's an optimization. This way if there are no elements or watermark events for a long time, we don't have to record every time the clock advanced forward.

Reply:
You would still be able to do that "compression" yourself and only add a single AdvanceProcessingTime yet still have the flexibility to record multiple AdvanceProcessingTime events if they were ever necessary.

Reply:
Turned into a single proto with the TestStreamPayload.Event.